src/python/pants/base/exception_sink.py

```diff
@@ -310,6 +338,18 @@ def trapped_signals(cls, new_signal_handler):
     finally:
       cls.reset_signal_handler(previous_signal_handler)
 
+  @classmethod
+  @contextmanager
+  def ignoring_sigint(cls):
```

Comment:
I think this might be more easily and explicitly done by creating a new `SignalHandler` subclass in `python_repl.py` with the lock and flag, then using `with ExceptionSink.trapped_signals(sigint_toggling_signal_handler):` there, instead of reaching in and mutating the global signal handler field, or moving signal handling concerns *into* `ExceptionSink` itself. This seems to be more testable, less coupled, and less spooky, and makes it clear when looking at the python repl task that some exclusive locking which modifies global state is going on.

This solution seems to work great as is -- my main goal with the global state in `ExceptionSink` was to make global state changes explicit and avoid mixing application-specific and process-handling logic. `SignalHandler` was an attempt to move task-specific or process-specific logic out of `ExceptionSink` and decouple it from the way we actually exit or log errors. Since it seems like `ExceptionSink.trapped_signals()` overwriting the existing signal handling provides the temporary override we want, it seems reasonable to use that approach instead of creating a new interface which reaches into the signal handler base class to modify it.

Reply:
See #7606 for an example of an `Exiter` accepting another instance of an `Exiter` and applying the prototype pattern in order to do a two-step teardown within `LocalPantsRunner` by calling another exiter passed into the constructor -- we could do something similar here with a `SignalHandler` subclass which does the same thing as a base signal handler instance, but also adds the `ignoring_sigint()` logic.

Reply:
I agree, wholeheartedly. But, in a world without `pantsd-runner`, the repl runs in a non-main thread, so when we try to do `ExceptionSink.trapped_signals()`, and this tries to `signal.signal()` for whatever handler we passed from `python-repl`, it will crash saying that this can only be done from the main thread.

Reply:
Oh! I didn't realize that, sorry. Could we then add an underscore to `SignalHandler._ignoring_sigint()` and then add something like:
> NB: This method calls `signal.signal()`, which will crash if not called from the main thread!

to the docstring for `.reset_signal_handler()` and `.trapped_signals()`?